Index: mod-sata.sh
===================================================================
--- mod-sata.sh	(wersja 12334)
+++ mod-sata.sh	(kopia robocza)
@@ -33,7 +33,7 @@
 	local PCI_ANY_ID=0x0000ffff pci_module vendor device subvendor subdevice class class_mask driver_data
 	while read pci_module vendor device subvendor subdevice class class_mask driver_data; do
 		# match class
-		[ "$(($req_class & $class_mask))" = "$class" ] || continue
+		[ "$(($req_class & $class_mask))" = "$(($class))" ] || continue
 		# match vendor
 		[ "$1" = "$vendor" ] || continue
 		# match device, allow PCI_ANY_ID
