Index: geninitrd
===================================================================
--- geninitrd	(wersja 12709)
+++ geninitrd	(wersja 12711)
@@ -395,6 +395,10 @@
 			# XXX: find a away to autodetect
 			smodule=$(basename_module $modpath)
 			case "$smodule" in
+				crc-t10dif)
+					find_depmod "-crct10dif-pclmul"
+					find_depmod "-crct10dif"
+					;;
 				libcrc32c)
 					find_depmod "-crc32c-intel"
 					find_depmod "-crc32c"
