Index: geninitrd
===================================================================
--- geninitrd	(wersja 8652)
+++ geninitrd	(kopia robocza)
@@ -258,7 +258,7 @@
 
 inst() {
 	if [ "$#" != "2" ];then
-		echo "usage: inst <file> <destination>"
+		echo >&2 "usage: inst <file> <destination>"
 		return
 	fi
 	debug "$1 -> $2"
@@ -1072,7 +1072,8 @@
 	mkdir -p $MNTIMAGE/etc
 	mkdir -p $MNTIMAGE/dev
 	resume_dev="$(awk '/^resume device =/ { print $4 } ' /etc/suspend.conf)"
-	inst /dev/snapshot $resume_dev $MNTIMAGE/dev
+	inst /dev/snapshot $MNTIMAGE/dev
+	inst $resume_dev $MNTIMAGE/dev
 	inst /etc/suspend.conf $MNTIMAGE/etc/suspend.conf
 	inst /usr/sbin/resume "$MNTIMAGE/bin/resume"
 	echo "resume" >> "$s"
