diff -ur geninitrd-8202.orig/geninitrd geninitrd-8202/geninitrd
--- geninitrd-8202.orig/geninitrd	2007-01-27 19:35:18.000000000 +0100
+++ geninitrd-8202/geninitrd	2007-02-05 11:31:33.000000000 +0100
@@ -934,12 +934,7 @@
 
 if [ "x" = "x$INITRDFS" ] ; then
 	if [ "x" = "x$FS" ] ; then
-		# default value
-		if [ "$pack_version" -ge "002005" ] ; then
-			INITRDFS="initramfs"
-		else
-			INITRDFS="rom"
-		fi
+		INITRDFS="rom"
 	else
 		echo >&2 "Warning: FS configuration options is obsoleted. Use INITRDFS instead"
 		INITRDFS="$FS"
@@ -1556,7 +1556,7 @@
 	init=/sbin/init
 fi
 umount /proc
-exec switch_root /newroot \$init
+exec chroot /newroot \$init
 EOF
 	# we need real file, not symlink
 	rm -f $MNTIMAGE/init
