Index: mod-md.sh
===================================================================
--- mod-md.sh	(wersja 11336)
+++ mod-md.sh	(kopia robocza)
@@ -10,9 +10,13 @@
 # @access	public
 setup_mod_md() {
 	mdassemble=$(find_tool $initrd_dir/mdassemble /sbin/initrd-mdassemble)
-	if [ ! -x /sbin/mdadm ] || [ ! -x "$mdassemble" ]; then
+	if [ ! -x "$mdassemble" ]; then
 		USE_MD=no
 	fi
+	mdadm=$(find_tool /sbin/mdadm $initrd_dir/mdadm /sbin/initrd-mdadm)
+	if [ ! -x "$mdadm" ]; then
+		USE_MD=no
+	fi
 }
 
 # return true if mdadm is set on $devpath
@@ -42,7 +46,7 @@
 
 	if [ -f /etc/mdadm.conf ]; then
 		debug "Finding RAID details using mdadm for rootdev=$1"
-		eval `(/sbin/mdadm -v --examine --scan --config=/etc/mdadm.conf 2> /dev/null;/sbin/mdadm -v --detail --scan --config=/etc/mdadm.conf 2> /dev/null) | awk -v rootdev="$1" '
+		eval `($mdadm -v --examine --scan --config=/etc/mdadm.conf 2> /dev/null;$mdadm -v --detail --scan --config=/etc/mdadm.conf 2> /dev/null) | awk -v rootdev="$1" '
 		BEGIN {
 			found = "no";
 			dev_list = "";
@@ -143,7 +147,7 @@
 
 		echo "DEVICE $cr_dev_list" >> "$DESTDIR/etc/mdadm.conf"
 		cr_dev_list_md="$(echo "$cr_dev_list" | xargs | awk ' { gsub(/ +/,",",$0); print $0; }')"
-		cr_md_conf=$(/sbin/mdadm --detail --brief --config=/etc/mdadm.conf $cr_rootdev | awk ' { gsub(/spares=[0-9]+/, "", $0); print $0; }')
+		cr_md_conf=$($mdadm --detail --brief --config=/etc/mdadm.conf $cr_rootdev | awk ' { gsub(/spares=[0-9]+/, "", $0); print $0; }')
 		if [ -n "$cr_md_conf" ]; then
 			echo "$cr_md_conf" >> "$DESTDIR/etc/mdadm.conf"
 		else
