Index: geninitrd
===================================================================
--- geninitrd	(wersja 12474)
+++ geninitrd	(kopia robocza)
@@ -792,8 +792,8 @@
 				if (c ~ "^" dev_pattern "$") partition = c;
 			}
 
-			$4 ~ partition { maj = $1; min = $2; }
-			$1 ~ maj && $2 ~ min { partition = $4; }
+			$4 == partition { maj = $1; min = $2; }
+			$1 == maj && $2 == min { partition = $4; }
 
 			END {
 				if (maj >= 0 && min >= 0) {
