Index: geninitrd
===================================================================
--- geninitrd	(.../tags/6894)	(revision 6997)
+++ geninitrd	(.../trunk)	(revision 6997)
@@ -355,7 +355,7 @@
 		getline x;
 		if (x ~ /devices=/) {
 			dev_list = x;
-			gsub("devices=", NUL, dev_list);
+			gsub(".*devices=", NUL, dev_list);
 			gsub(",", " ", dev_list);
 		}
 	    }
@@ -1197,6 +1197,8 @@
 		mkdir -p $MNTIMAGE/dev/mapper
 		mknod $MNTIMAGE/dev/mapper/control c 10 63
 		for device in $PVDEVICES; do
+			# if LVM on RAID then device might be copied already in gen_softraid
+			[ -e "$MNTIMAGE/dev/$(basename $device)" ] && continue
 			cp -HR $device $MNTIMAGE/dev/
 		done
 	fi
