diff -ur geninitrd-6711-o/geninitrd geninitrd-6711/geninitrd
--- geninitrd-6711-o/geninitrd	2005-12-26 09:14:10.000000000 -0700
+++ geninitrd-6711/geninitrd	2005-12-26 20:13:27.000000000 -0700
@@ -673,7 +673,7 @@
 	        module="`echo $mod | awk -F/ '{ $0=$NF } /'$modext'.*$/ { gsub(/'$modext'.*/, NIL, $0); } { print $0; }'`"
 	        options="`awk '{ if($1 == "options" && $2 == "'${module}'") { for(i=3;i<=NF;i++) printf("%s ",$i); }}' "$modulefile"`"
 
-		sleep_module=$(echo "${module}" | awk ' { gsub("-", "_", $0 } { print $0; } ')
+		sleep_module=$(echo "${module}" | awk ' { gsub("-", "_", $0) } { print $0; } ')
 		sleep_var="$(eval echo \$MODULE_${sleep_module}_USLEEP)"
 
 	        if [ -n "$verbose" ]; then
@@ -1242,7 +1242,7 @@
 		mkdir -p $MNTIMAGE/dev/mapper
 		mknod $MNTIMAGE/dev/mapper/control c 10 63
 		for device in $PVDEVICES; do
-			cp -HR $device $MNTIMAGE
+			cp -HR $device $MNTIMAGE/dev
 		done
 	fi
 	echo "mount -t proc none /proc" >> "$s"
