--- geninitrd-8985/geninitrd~	2007-11-04 22:52:56.894367876 +0200
+++ geninitrd-8985/geninitrd	2007-11-04 22:53:39.836233209 +0200
@@ -1148,12 +1148,7 @@
 
 if [ -z "$INITRDFS" ]; then
 	if [ -z = "$FS" ]; then
-		# default value
-		if [ "$pack_version" -ge "002005" ]; then
-			INITRDFS="initramfs"
-		else
-			INITRDFS="rom"
-		fi
+		INITRDFS="rom"
 	else
 		echo >&2 "Warning: FS configuration options is obsoleted. Use INITRDFS instead"
 		INITRDFS="$FS"
