diff -urN geninitrd-4605.org/geninitrd geninitrd-4605/geninitrd
--- geninitrd-4605.org/geninitrd	2004-09-25 23:18:06.000000000 +0200
+++ geninitrd-4605/geninitrd	2005-06-27 08:29:53.000000000 +0200
@@ -322,7 +322,7 @@
 find_modules_softraid() {
         if [ -f /etc/mdadm.conf ] ; then
 	    [ -n "$verbose" ] && echo "Finding RAID details using mdadm for rootdev=$1"
-	    eval `/sbin/mdadm --examine  --scan --config=/etc/mdadm.conf | \
+	    eval `/sbin/mdadm -v --examine  --scan --config=/etc/mdadm.conf | \
 	    	  awk -v rootdev="$1" '
 	BEGIN {
 		found = "no";
