--- ./geninitrd	2005-10-13 18:56:09.000000000 +0300
+++ /tmp/geninitrd	2005-10-13 18:56:02.000000000 +0300
@@ -1039,6 +1039,42 @@
 EOF
 }
 
+initrd_gen_udev() {
+	mkdir -p "$MNTIMAGE/sbin"
+	inst /sbin/start_udev $MNTIMAGE/sbin/start_udev
+cat <<-EOF >> "$s"
+	/sbin/start_udev
+EOF
+}
+
+initrd_gen_dmraid() {
+	if [ ! -x /sbin/dmraid-initrd ] ; then
+		echo "/sbin/dmraid-initrd is missing missing !"
+		exit 1
+	fi
+
+	mkdir "$MNTIMAGE/sbin"
+#	cp -a /dev/mapper "$MNTIMAGE/dev/mapper"
+	mkdir -p "$MNTIMAGE/dev/mapper"
+	mknod $MNTIMAGE/dev/sda b 8 0
+	mknod $MNTIMAGE/dev/sdb b 8 16
+	mkdir -p "$MNTIMAGE/var/lock/dmraid"
+	inst /sbin/dmraid-initrd $MNTIMAGE/sbin/dmraid
+	# need dmraid > 1.0.0-rc9, or otherwise uncomment /var/lock/dmraid tmpfs mount and -i option from dmraid
+cat <<-EOF >> "$s"
+	mount -t proc none /proc
+	mount -t sysfs none /sys
+	mount -o mode=0755 -t tmpfs none /dev/mapper
+#	mount -o mode=0755 -t tmpfs none /var/lock/dmraid
+	/sbin/dmraid -ay -i
+#	sh
+#	umount /dev/mapper
+#	umount /var/lock/dmraid
+	umount /sys
+	umount /proc
+EOF
+}
+
 initrd_gen_softraid() {
 	[ -n "$verbose" ] && echo "Setting up mdadm..."
 	
@@ -1210,6 +1247,9 @@
 	initrd_gen_suspend
 fi
 
+#initrd_gen_udev
+initrd_gen_dmraid
+
 if is_yes "$usenfs" ; then
 	initrd_gen_nfs
 elif is_yes "$USERAIDSTART" && is_yes "$raidfound" ; then
