Index: geninitrd
===================================================================
--- geninitrd	(wersja 9989)
+++ geninitrd	(kopia robocza)
@@ -1135,11 +1135,9 @@
 	EOF
 
 	inst_exec /sbin/initrd-udevadm /sbin/udevadm
-	ln -s udevadm $DESTDIR/sbin/udevsettle
-	ln -s udevadm $DESTDIR/sbin/udevtrigger
 	add_linuxrc <<-'EOF'
-		/sbin/udevtrigger
-		/sbin/udevsettle
+		/sbin/udevadm trigger
+		/sbin/udevadm settle
 	EOF
 
 	busybox_applet killall
