diff --git a/geninitrd b/geninitrd
index 7b962b5..a4de196 100755
--- a/geninitrd
+++ b/geninitrd
@@ -1103,8 +1103,7 @@ sym_exists() {
 # find best compressor (or forced one) for initrd
 find_compressor() {
 	local mode="$1"
-	# fastest initrd decompression speed is first
-	local compressors='lzo gzip xz lzma bzip2'
+	local compressors='xz lzma bzip2 gzip lzo'
 
 	# a specified one, take it
 	if ! is_yes "$mode"; then
diff --git a/mod-md.sh b/mod-md.sh
index a0cd7aa..d9093d4 100644
--- a/mod-md.sh
+++ b/mod-md.sh
@@ -139,13 +139,15 @@ initrd_gen_md() {
 
 	inst_exec $mdassemble /bin/mdassemble
 
+	echo "DEVICE partitions containers" >> "$DESTDIR/etc/mdadm.conf"
+
 	# LVM on RAID case
 	local dev_list_extra ex_dev
 	dev_list_extra=$(awk '/^DEVICE / { for (i=2; i<=NF; i++) { printf "%s ", $i; }; } ' /etc/mdadm.conf | xargs)
 	new_dev_list_extra=""
 	for ex_dev in $dev_list_extra; do
-		if [ "$ex_dev" = "partitions" ]; then
-			echo "DEVICE partitions" >> "$DESTDIR/etc/mdadm.conf"
+		if [ "$ex_dev" = "partitions" -o "$ex_dev" = "containers" ]; then
+			echo "DEVICE $ex_dev" >> "$DESTDIR/etc/mdadm.conf"
 			# FIXME: find and copy partition devices from /proc/partitions
 			#        - best if done at runtime, now initrd gen time
 			continue
commit 11202855dbaee5bc32209913df1ae8738c210c39
Author: Arkadiusz Miśkiewicz <arekm@maven.pl>
Date:   Sat May 14 00:36:02 2016 +0200

    Make rootdev fallback work.

diff --git a/geninitrd b/geninitrd
index 92a519d..771245e 100755
--- a/geninitrd
+++ b/geninitrd
@@ -1004,7 +1004,11 @@ initrd_gen_initramfs_switchroot() {
 
 		# XXX hack, fallback to rootdev from geninitrd time
 		if [ ! -e "$device" ]; then
+	EOF
+	add_linuxrc <<-EOF
 			device="$rootdev"
+	EOF
+	add_linuxrc <<-'EOF'
 			echo "DEVICE set to $device based on fstab entry from initrd gen time"
 		fi
 
