Index: geninitrd
===================================================================
--- geninitrd	(wersja 12174)
+++ geninitrd	(kopia robocza)
@@ -772,9 +772,9 @@
 			device=$rootdev
 		fi
 
-		[ -n "$ROOTFSFLAG" ] && ROOTFSFLAGS="-o $ROOTFSFLAGS"
+		[ -n "$ROOTFSFLAGS" ] && ROOTFSFLAGS="-o $ROOTFSFLAGS"
 
-		mount -t $rootfs -r $device $ROOTFSFLAGS /newroot
+		mount -t $rootfs -r $device $ROOTFSFLAGS /newroot || echo "Mount of rootfs failed."
 		init="$(echo "$CMDLINE" | busybox awk '/init=\// { gsub(/.*init=/,NIL,$0); gsub(/ .*/,NIL,$0); print }')"
 		if [ -z "$init" -o ! -x "/newroot$init" ]; then
 			init=/sbin/init
Index: geninitrd
===================================================================
--- geninitrd	(wersja 12182)
+++ geninitrd	(kopia robocza)
@@ -709,7 +709,7 @@
 			rootnr="$(busybox awk -v rootnode="${ROOT##/dev/}" '$4 == rootnode { print 256 * $1 + $2 }' /proc/partitions)"
 			# fallback to ls
 			if [ -z "$rootnr" ]; then
-				rootnr="$(busybox ls -lL ${ROOT} | awk '{if (/^b/) { print 256 * $3 + $4; }}')"
+				rootnr="$(busybox ls -lL ${ROOT} | busybox awk '{if (/^b/) { print 256 * $3 + $4; }}')"
 			fi
 			if [ -n "$rootnr" ]; then
 				echo "$rootnr" > /proc/sys/kernel/real-root-dev
